indexing-service/src/main/java/org/apache/druid/indexing/overlord/PortFinder.java

```diff
@@ -55,39 +62,63 @@ private static boolean canBind(int portNum)
 
   public synchronized int findUnusedPort()
   {
-    int port = chooseNext(startPort);
-    while (!canBind(port)) {
-      port = chooseNext(port + 1);
+    if (candidatePorts != null && !candidatePorts.isEmpty()) {
+      int port = chooseFromCandidates();
+      usedPorts.add(port);
+      return port;
+    } else {
+      int port = chooseNext(startPort);
+      while (!canBind(port)) {
+        port = chooseNext(port + 1);
+      }
+      usedPorts.add(port);
+      return port;
     }
-    usedPorts.add(port);
-    return port;
   }
 
-  public synchronized Pair<Integer, Integer> findTwoConsecutiveUnusedPorts()
+  public synchronized Pair<Integer, Integer> findTwoUnusedPorts()
   {
-    int firstPort = chooseNext(startPort);
-    while (!canBind(firstPort) || !canBind(firstPort + 1)) {
-      firstPort = chooseNext(firstPort + 1);
+    if (candidatePorts != null && !candidatePorts.isEmpty()) {
+      int firstPort = chooseFromCandidates();
+      int secondPort = chooseFromCandidates();
+      usedPorts.add(firstPort);
+      usedPorts.add(secondPort);
+      return new Pair<>(firstPort, secondPort);
```

Comment:
if we are not returning consecutive ports then why the impl can just be calling the `findUnusedPort()` twice ?

Reply:
Fixed.